presto-main/src/main/java/io/prestosql/sql/planner/DesugarRowSubscriptRewriter.java

```diff
@@ -0,0 +1,114 @@
+import io.prestosql.sql.tree.LongLiteral;
+import io.prestosql.sql.tree.NodeRef;
+import io.prestosql.sql.tree.SubscriptExpression;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Replaces subscript expression on Row
+ * with cast and dereference:
+ * <pre>
+ *     ROW (1, 'a', 2) [2]
+ * </pre>
+ * is transformed into:
+ * <pre>
+ *     (CAST (ROW (1, 'a', 2) AS ROW (field_0 bigint, field_1 varchar(1), field_2 bigint))).field_1
+ * </pre>
+ */
+public class DesugarRowSubscriptRewriter
+{
+    private DesugarRowSubscriptRewriter() {}
+
+    public static Expression rewrite(Expression expression, Session session, TypeAnalyzer typeAnalyzer, SymbolAllocator symbolAllocator)
+    {
+        requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+
+        /*if (expression instanceof SymbolReference) {
+            return expression;
+        }*/
+        Map<NodeRef<Expression>, Type> expressionTypes = typeAnalyzer.getTypes(session, symbolAllocator.getTypes(), expression);
+
+        return rewrite(expression, expressionTypes);
+    }
+
+    public static Expression rewrite(Expression expression, Map<NodeRef<Expression>, Type> expressionTypes)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new Visitor(expressionTypes), expression);
+    }
+
+    private static class Visitor
+            extends ExpressionRewriter<Void>
+    {
+        private final Map<NodeRef<Expression>, Type> expressionTypes;
+
+        public Visitor(Map<NodeRef<Expression>, Type> expressionTypes)
+        {
+            this.expressionTypes = ImmutableMap.copyOf(requireNonNull(expressionTypes, "expressionTypes is null"));
+        }
+
+        @Override
+        public Expression rewriteSubscriptExpression(SubscriptExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            Expression base = treeRewriter.rewrite(node.getBase(), context);
+            Expression index = treeRewriter.rewrite(node.getIndex(), context);
+
+            Type type = getType(base);
+            if (!(type instanceof RowType)) {
+                return new SubscriptExpression(base, index);
+            }
+
+            RowType rowType = (RowType) type;
+            int position = toIntExact(((LongLiteral) index).getValue() - 1);
+            Optional<String> fieldName = rowType.getFields().get(position).getName();
+
+            // Do not cast if Row fields are named
+            if (fieldName.isPresent()) {
+                return new DereferenceExpression(base, new Identifier(fieldName.get()));
+            }
+
+            // Cast to Row with named fields
+            ImmutableList.Builder<RowType.Field> namedFields = new ImmutableList.Builder<>();
+            for (int i = 0; i < rowType.getFields().size(); i++) {
+                namedFields.add(new RowType.Field(Optional.of(String.format("field_%s", i)), rowType.getTypeParameters().get(i)));
```

Comment:
Maybe use `f%s` as the field names. It's more concise and it's easier to read the plans. For the example above (`x[1][1]`):

```
CAST(CAST("field" AS row(field_0 row(integer,integer))).field_0 AS row(field_0 integer,field_1 integer)).field_0
```
vs
```
CAST(CAST("field" AS row(f0 row(integer,integer))).f0 AS row(f0 integer,f1 integer)).f0
```